examples/c-comparisons/script.jl

```diff
@@ -0,0 +1,125 @@
+
+# ## Generate some training data
+#
+# We create a binary-labelled toy dataset:
+
+Xgrid = -4:0.1:29  # for visualization
+X = range(0, 23.5; length=48)  # training inputs
+f(x) = 3 * sin(10 + 0.6x) + sin(0.1x) - 1  # latent function
+fs = f.(X)  # latent function values at training inputs
+const invlink = logistic  # could use other invlink, e.g. normcdf(f) = cdf(Normal(), f)
+ps = invlink.(fs)  # probabilities at the training inputs
+Y = [rand(Bernoulli(p)) for p in ps]  # observations at the training inputs
+
+function plot_data()
+    plot(; xlims=extrema(Xgrid), xticks=0:6:24)
+    plot!(Xgrid, invlink ∘ f; label="true probabilities")
+    return scatter!(X, Y; label="observations", color=3)
+end
+
+plot_data()
+
+# ## Creating the latent GP
+# 
+# Here we write a function that creates our latent GP prior, given the
+# hyperparameter vector `theta`. Compared to a "vanilla" GP, the `LatentGP`
+# requires a function or functor that maps from the latent GP `f` to the
+# distribution of observations `y`. This functor is commonly called "the
+# likelihood".
+
+function build_latent_gp(theta)
+    # `theta` is unconstrained, but kernel variance and lengthscale must be positive:
+    variance = softplus(theta[1])
+    lengthscale = softplus(theta[2])
+
+    kernel = variance * with_lengthscale(SqExponentialKernel(), lengthscale)
+
+    dist_y_given_f = BernoulliLikelihood()  # has logistic invlink by default
+    # We could also be explicit and define it as a function:
+    # dist_y_given_f(f) = Bernoulli(invlink(f))
+
+    jitter = 1e-8  # required for numeric stability [TODO: where to explain this better?]
+    return LatentGP(GP(kernel), dist_y_given_f, jitter)
+end
+
+# We define a latent GP at our initial hyperparameter values, here with
+# variance 1.0 and lengthscale 5.0:
+
+theta0 = [invsoftplus(1.0), invsoftplus(5.0)]
+
+lf = build_latent_gp(theta0)
+
+lf.f.kernel
+
+# We can now compute the posterior `p(f | y)` under the Laplace approximation:
+
+f_post = posterior(LaplaceApproximation(), lf(X), Y)
+
+# Let's plot samples from this approximate posterior:
+
+function plot_samples!(Xgrid, fpost; samples=100, color=2)
+    fx = fpost(Xgrid, 1e-8)
+    fsamples = rand(fx, samples)
+    plot!(Xgrid, invlink.(fsamples); color, alpha=0.2, label="")
+    return plot!(Xgrid, invlink.(mean(fx)); color, lw=2, label="posterior fit")
+end
+
+p1 = plot_data()
+plot_samples!(Xgrid, f_post)
+
+# We can improve this fit by optimising the hyperparameters.
+
+# ## Optimise the parameters
+#
+# ApproximateGPs provides a convenience function `build_laplace_objective` that
+# constructs an objective function for optimising the hyperparameters, based on
+# the Laplace approximation to the log marginal likelihood.
+# We pass this objective to Optim.jl's LBFGS optimiser:
+
+objective = build_laplace_objective(build_latent_gp, X, Y)
+
+training_results = Optim.optimize(
+    objective, θ -> only(Zygote.gradient(objective, θ)), theta0, LBFGS(); inplace=false
```

Comment:
The  `only` syntax looks a bit confusing, can't we pass the objective function directly to Optim?

Reply:
What sort of thing did you have in mind Theo? Would doing this kind of thing not mandate that ApproximateGPs know about Zygote or something?

Reply:
I mean that people might expect to just pass the objective alone without having to pass the gradient itself like
`optimize(objective, theta0, LBFGS(); ...)`, eventually telling Optim to use Zygote (I don't know if it is possible)

Reply:
`only` is because Zygote returns a one-tuple, but Optim requires the actual gradient. And Optim doesn't know anything about Zygote - you have to explicitly provide the gradient function.